src/cpp/rtps/reader/WriterProxy.cpp

```diff
@@ -206,20 +206,36 @@ void WriterProxy::missing_changes_update(
     }
 }
 
-void WriterProxy::lost_changes_update(
+int32_t WriterProxy::lost_changes_update(
         const SequenceNumber_t& seq_num)
 {
 #ifdef SHOULD_DEBUG_LINUX
     assert(get_mutex_owner() == get_thread_id());
 #endif // SHOULD_DEBUG_LINUX
 
     logInfo(RTPS_READER, guid().entityId << ": up to seq_num: " << seq_num);
+    int32_t current_sample_lost = 0;
 
     // Check was not removed from container.
-    if (seq_num > changes_from_writer_low_mark_)
+    if (seq_num > (changes_from_writer_low_mark_ + 1))
     {
         // Remove all received changes with a sequence lower than seq_num
         ChangeIterator it = std::lower_bound(changes_received_.begin(), changes_received_.end(), seq_num);
+        if (!changes_received_.empty())
+        {
+            current_sample_lost =
+                    static_cast<int32_t>((*changes_received_.begin()).to64long() -
+                    (changes_from_writer_low_mark_.to64long() + 1));
+            auto distance = std::distance(changes_received_.begin(), it);
+            current_sample_lost +=
+                    static_cast<int32_t>(seq_num.to64long() - (*changes_received_.begin()).to64long() - distance);
+        }
+        else
+        {
+            current_sample_lost =
+                    static_cast<int32_t>(seq_num.to64long() -
+                    (changes_from_writer_low_mark_.to64long() + 1));
+        }
```

Comment:
Beware of arithmetic overflows. What should happen if more than 2^31 samples are lost?

Reply:
Done in dead6d9e7